### ISSUE TITLE
Incorect SHA256

### ISSUE THREAD

**Issue by chrisburr** [regro/cf-scripts#500]
The update pull request (https://github.com/conda-forge/geant4-feedstock/pull/1) generated an incorrect SHA256. The hash used by the bot does however match with what I get locally using `wget`:

```bash
$ wget https://geant4-data.web.cern.ch/geant4-data/releases/geant4.10.05.01.tar.gz
$ openssl sha256 geant4.10.05.p01.tar.gz
SHA256(geant4.10.05.p01.tar.gz)= f4a292220500fad17e0167ce3153e96e3410ecbe96284e572dc707f63523bdff
```

I'm not sure why `conda-build` is getting something different, any ideas?

**Comment by msarahan** [comment]
Every so often, people have misconfigured webservers that don't express compression settings correctly.  Things end up recompressed or decompressed when we're expecting them raw.  Try downloading the file locally and decompressing (but not extracting) the .tar.gz.  Is that the alternate SHA256 you see?

I think the answer in the past has been to just adjust to what conda-build sees, assuming we understand the reason.  It would be nice to fix this, but fortunately(?) it's a very rare problem and hasn't been a priority.  IIRC, we saw this before with the bzip2 source - maybe you can find more info by reading the history on that feedstock or searching for related stuff.

**Comment by chrisburr** [comment]
Yup that's the reason:

```bash
$ wget https://geant4-data.web.cern.ch/geant4-data/releases/geant4.10.05.p01.tar.gz
$ gzip -d geant4.10.05.p01.tar.gz --stdout > geant4.10.05.p01.tar
$ openssl sha256 *
SHA256(geant4.10.05.p01.tar)= c6bc5a913b846b9fe7dd6e6a8bf11ef315e00a37c14e67b0a5213205a9abe79f
SHA256(geant4.10.05.p01.tar.gz)= f4a292220500fad17e0167ce3153e96e3410ecbe96284e572dc707f63523bdff
```

The first is the hash `conda-build` expects and the second is the hash `cf-scripts` is generating.

It makes sense for it to not be a priority and it's reassuring to know why it doesn't match, thanks!

**Comment by CJ-Wright** [comment]
Can we close this?

**Comment by chrisburr** [comment]
It’s still a problem for feedstocks with geant4 in the name (and possibly others).

**Comment by CJ-Wright** [comment]
Ah ok, this seems more on the conda build side?

**Comment by chrisburr** [comment]
**TLDR:** `cf-scripts` is using `urllib` to download files whereas `conda` uses `requests` and this can introduce differences in what bytes are hashed 🤦‍♂

**Full details:** This became a bigger issue elsewhere so I've dug deeper into the problem. The issue is that the server incorrectly sends the header `Content-Encoding: gzip` despite a second level of gzip compression not being applied. This can be reproduced locally by modifying the Python HTTP server to always send `Content-Encoding: gzip`.

The inconsistency here is that [`conda` uses `requests`](https://github.com/conda/conda/blob/beb6e8a957dbcf4e4fc325b2bc2f8b6ced5ef8b9/conda/gateways/connection/download.py#L87) to download the file, while  `cf-scripts` uses `urllib` ([via `rever`](https://github.com/regro/rever/blob/1091f542f581140b9451ec8c6dd0fd4c2c5f4f2f/rever/tools.xsh#L179)). Some libraries like `wget`, `curl` and `urllib` somehow manage to realise that the header is nonsense, whereas `requests` respects what the header says.

### Stand alone reproducer

```python
import urllib
import requests

url = 'http://geant4-data.web.cern.ch/geant4-data/releases/geant4.10.06.tar.gz'

print('Plain curl', end='\n    ')
!curl --silent -L $url | openssl sha256
print('Curl with added gzip decompression', end='\n    ')
!curl --silent -L $url | gzip -d | openssl sha256

print()

resp = requests.get(url)
print('With requests', end='\n    ')
with open('/tmp/test.tar.gz', 'wb') as fh:
    fh.write(resp.content)
!cat /tmp/test.tar.gz | openssl sha256

resp = urllib.request.urlopen(url)
print('With urllib', end='\n    ')
with open('/tmp/test.tar.gz', 'wb') as fh:
    fh.write(resp.read())
!cat /tmp/test.tar.gz | openssl sha256
```

### Output of reproducer

```
Plain curl
    1424c5a0e37adf577f265984956a77b19701643324e87568c5cb69adc59e3199
Curl with added gzip decompression
    dc9adfe6aa508c6944e8a5ca14a6a799e36408d5784757aca1f2c51819fc467d

With requests
    dc9adfe6aa508c6944e8a5ca14a6a799e36408d5784757aca1f2c51819fc467d
With urllib
    1424c5a0e37adf577f265984956a77b19701643324e87568c5cb69adc59e3199
```
